### ISSUE TITLE
Network Contributor role assignment fails with UserAssigned Identity and Azure subnet

### ISSUE THREAD

**Issue by bonddim** [Azure/terraform-azurerm-aks#405]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Greenfield/Brownfield provisioning

greenfield

### Terraform Version

1.5.2

### Module Version

7.1.0

### AzureRM Provider Version

3.63.0

### Affected Resource(s)/Data Source(s)

azurerm_role_assignment

### Terraform Configuration Files

```hcl
data "azurerm_subnet" "main" {
  name                 = "cluster"
  virtual_network_name = var.virtual_network_name
  resource_group_name  = var.virtual_network_group_name
}

resource "azurerm_resource_group" "main" {
  name     = "rg-laki-shared"
  location = var.location
}

resource "azurerm_user_assigned_identity" "main" {
  resource_group_name = azurerm_resource_group.main.name
  location            = azurerm_resource_group.main.location

  name = "uami-${var.kubernetes_cluster_name}"
}

module "aks" {
  source  = "Azure/aks/azurerm"
  version = "7.1.0"

  cluster_name        = var.kubernetes_cluster_name
  prefix              = var.kubernetes_cluster_name
  resource_group_name = azurerm_resource_group.main.name
  
  identity_ids              = [azurerm_user_assigned_identity.main.id]
  identity_type             = "UserAssigned"
  
  vnet_subnet_id = data.azurerm_subnet.main.id

  create_role_assignment_network_contributor = true
}
```

### tfvars variables values

```hcl
location                   = "northeurope"
kubernetes_cluster_name    = "myAks"
virtual_network_name       = "myVnet"
virtual_network_group_name = "myRg"
```

### Debug Output/Panic Output

```shell
│ Error: authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InvalidPrincipalId" Message="A valid principal ID must be provided for role assignment."
```

### Expected Behaviour

Assign Network Contributor role to provided UserAssigned Identity Id

### Actual Behaviour

Module tries to assign role to non-existent SystemAssigned Identity

### Steps to Reproduce

`terraform apply`

**Comment by zioproto** [comment]
@lonegunmanb I was trying to reproduce this issue and on `terraform plan` I got:

```
│ Error: Invalid for_each argument
│ 
│   on .terraform/modules/aks/main.tf line 749, in resource "azurerm_role_assignment" "network_contributor":
│  749:   for_each = var.create_role_assignment_network_contributor ? local.subnet_ids : []
│     ├────────────────
│     │ local.subnet_ids is a set of dynamic, known only after apply
│     │ var.create_role_assignment_network_contributor is true
│ 
│ The "for_each" set includes values derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys
│ that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map value where the keys are defined statically in your configuration and where only the values
│ contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply only the resources that the for_each value depends on, and then apply a second time to fully
│ converge.
```

I believe @bonddim does not hit this issue because in the reproduction code provided the subnet is a datasource and not a resource.

Is possible this is a regression coming from PR #374 ?

**Comment by bonddim** [comment]
@zioproto, correct. 
I got the same error when trying to use `azurerm_subnet` resource. 
But anyway, opened issue is not related to `local.subnet_ids` calculations. 
The most obvious fix for current issue is to add variable for `principal_id` and use it conditionally for role assignment.

**Comment by lonegunmanb** [comment]
Thanks @bonddim for opening this issue, I've debugged the module, The error was caused by the following code:

```hcl
resource "azurerm_role_assignment" "network_contributor" {
  for_each = var.create_role_assignment_network_contributor ? local.subnet_ids : []

  principal_id         = azurerm_kubernetes_cluster.main.identity[0].principal_id
  scope                = each.value
  role_definition_name = "Network Contributor"
}
```

It seems like once we use user assigned identity, its principal id won't be exported to aks cluster's identity nested block. We need an explicit user assigned identity datasource to retrieve the principal id.

I'll try a hotfix, once it pass the test I'll release a patch version to fix this issue.

Meanwhile, we need a separate issue to trach the `for_each` issue.
